Modules/event.py

```diff
@@ -0,0 +1,157 @@
+All rights reserved.
+
+Licensed under the BSD 3-Clause License.
+
+See LICENSE.md
+"""
+import logging
+from functools import partial
+from typing import Callable, List, Dict, Union
+
+# typedef
+subscriptions = List[Callable]
+log = logging.getLogger(f"mecha.{__name__}")
+
+
+class Event:
+    """
+    Registers a async function as an event.
+
+    The registered event will be the name of the decorated function, when used in the @Event form.
+
+    If used in the Event(name) form, the event's name will be the `coro` parameter.
+
+    Examples:
+        any async function definition can be decorated, and the function will be registered as a
+        command
+         >>> @Event
+         ... async def my_event(*args, **kwargs):
+         ...    pass
+         >>> "my_event" in Event.events
+         True
+
+         an event can also be defined without a function definition
+         >>> my_other_event = Event("my_other_event")
+         >>> "my_other_event" in Event.events
+         True
+
+         Notes:
+             - Events declared via the Event("name") form will need to be invoked by something.
+
+             - Events are callables, and should be treated like async functions when called.
+
+             - If used in the @Event form, the decorated function will be called before any
+             subscribers.
+
+             - the return value of decorated events are discarded.
+    """
+    events: Dict[str, 'Event'] = {}
+    """Registry mapping event names to their subscribers"""
+
+    def __init__(self, coro: Union[Callable, str]):
+        log.debug(f"decorating coro {coro}")
+        if callable(coro):
+            # decorator form
+            self.decorated_coro = coro
+            self.name = coro.__name__
+        elif isinstance(coro, str):
+            # Event("name") form
+            self.decorated_coro = None
+            self.name = coro
+        else:
+            raise TypeError(f"expected `coro` to be either a callable or a string. "
+                            f"got type {type(coro)}")
+
+        self.subscribers: subscriptions = []
+        """This events subscribers"""
+
+        Event._register(self)
+
+    async def _emit(self, *args, **kwargs):
+        """Emit the event to all subscribers"""
+        for subscriber in self.subscribers:
+            log.debug(f"calling subscriber {subscriber}...")
+            await subscriber(*args, **kwargs)
+
+    async def __call__(self, *args, **kwargs):
+        log.debug(f"event {self.name} invoked...")
+        if self.decorated_coro is not None:
+            await self.decorated_coro(*args, **kwargs)
+        log.debug(f"emitting event to subscribers...")
+        await self._emit(*args, **kwargs)
+
+    def __get__(self, instance, owner):
+        # i honestly have no idea what black magic this is.
+        # https://stackoverflow.com/questions/30104047/how-can-i-decorate-an-instance-method-with-a-decorator-class
+        return partial(self.__call__, instance)
```

Comment:
Let's have an example on how this 'black magic' works:
```python
class Bubble(object):
  def bath(self): ...

  @Event
  def gum(self): ...

my_bubble = Bubble()
```
When you call `Bubble.bath`, you have to provide the `self` argument yourself, as expected. You are calling a function like any other, `self` has no special significance here.

### Bound Methods ###
When, however, you access `my_bubble.bath`, you aren't getting a function. You are getting a **bound method**. The bound method behaves similarly to the function, with one key difference: The first argument is set to `my_bubble`. You can actually see what it was bound to by getting `my_bubble.bath.__self__`.
In essence, a method acts as follows:
```python
class MethodType(object):
  ...
  def __call__(self, *args, **kwargs):
    return self.__func__(self.__self__, *args, **kwargs)
  ...
```
In our example, `__self__` would be `my_bubble` and `__func__` would be `Bubble.bath`.

### The Problem ###
The problem arises for `Bubble.gum`.
Here, the function `gum` is replaced by an `Event` object, so the class `Bubble` now has an attribute `gum` of type `Event`. This means that no binding will take place. `Bubble.gum is my_bubble.gum` evaluates to True. The decorated function, however, is expecting a traditional `self` argument, i.e. the instance it is called on.

Your solution is to make `Event` a delegate. When `my_bubble.gum` is accessed, `my_bubble.gum.__get__` is called with `instance=my_bubble` and `owner=Bubble`. In `__get__`, you use `partial` to manually bind the first argument of `Event.__call__` to the instance `gum` was accessed on, i.e. `my_bubble` and returns the result. This means that `my_bubble.gum(7)` causes the underlying function to be called with `(my_bubble, 7)`.

### But why so complicated? ###
This was the happy path, and it works fine. Now, let's look at calling `Bubble.gum()` directly on the class. `__get__` will still be called, only with `instance=None`. This breaks the usual idea of being able to specify your own `self` argument, as it will be bound to None. You'll have to check for that and handle it.

Oh, and class as well as static methods won't work at all, just as a note.

This all strikes me as massively overcomplicated. Why should events be created by decorator at all? Why should it be called to trigger the event, that is simply misleading. Let me suggest three arguably simpler APIs:
```python
>>> bob = Event("bob")
... @bob.subscribe
... def bob1():
...   print("bob1")
... @Event("bob").subscribe
... def bob2():
...   print("bob2")
... bob.emit()
bob1
bob2
```
Or, without any classes:
```python
>>> @subscribe("bob")
... def bob1():
...   print("bob1")
... @subscribe("bob")
... def bob2():
...   print("bob2")
... emit("bob")
bob1
bob2
```
Or why allow arbitrary events at all, won't that only lead to typos? How about an enum?
```python
>>> @subscribe(Event.BOB)
... def bob1():
...   print("bob1")
... @subscribe(Event.BOB)
... def bob2():
...   print("https://youtu.be/FU-tuY0Z7nQ")
... emit(Event.BOB)  # Or Event.BOB.emit() ?
bob1
https://youtu.be/FU-tuY0Z7nQ
```